### ISSUE TITLE
GCD slower than hoped in ZZ[x]

### ISSUE THREAD

**Issue by JohnAAbbott** [oscar-system/Oscar.jl#2242]
**Describe the bug**
Below is some Julia/Oscar code to generate a pair of polynomials in ZZ[x] whose gcd is just x, but it takes surprisingly long to compute the answer.  UPB is a parameter for the family of polynomial pairs.  With UPB=100 total computation time is about 100s on my computer; with UPB=200 total computation time is 2+ hours (maybe even 10000s?)

**To Reproduce**
Steps to reproduce the behavior, please provide a code snippet that triggers
the bug.
```
using Oscar
# SLOW GCD:  a family of polynomial pairs parametrized by UPB (positive integer)
#            gcd of each pair is x, but comnputation time is a bit long

# >>> IMPORTANT <<<   the nested "for" loops can take a while to create f & g

P,x = polynomial_ring(ZZ,"x");

UPB = 100;

f = x;  #  starting with f=1 and g=1 then it takes only about 5s (UPB=200)
g = x;  #  starting with f=x and g=x then it takes about 2000s   (UPB=200)
for n in 1:UPB
  for d in 1:UPB
    if (gcd(d,n) == 1)
      f = f * (d*x-n);
      g = g * (d*x^2+n);
    end #if
  end #for
end #for

println(degree(f));
println(degree(g));

@time h = gcd(f,g);
h  # gcd is just x
```

**Expected behavior**
Can it be made faster?   Computing `gcd(f/x,g/x)` is much faster (because gcd=1, so the algorithm can short-cut).

**System (please complete the following information):**
Please paste the output of `Oscar.versioninfo(full=true)` below. If this does
not work, please paste the output of Julia's `versioninfo()` and your Oscar
version.
```
julia> Oscar.versioninfo(full=true)
OSCAR version 0.12.0-DEV - #JAA/exterior_algebra, fffc2c2e -- 2023-04-11 14:47:24 +0200
  combining:
    AbstractAlgebra.jl   v0.29.2
    GAP.jl               v0.9.4
    Hecke.jl             v0.18.8
    Nemo.jl              v0.33.7
    Polymake.jl          v0.9.0
    Singular.jl          v0.18.1
  building on:
    Antic_jll               v0.201.500+0
    Arb_jll                 v200.2300.0+0
    Calcium_jll             v0.401.100+0
    FLINT_jll               v200.900.5+0
    GAP_jll                 v400.1200.200+0
    Singular_jll            v403.101.500+0
    libpolymake_julia_jll   v0.9.0+0
    libsingular_julia_jll   v0.29.0+0
    polymake_jll            v400.900.0+0
See `]st -m` for a full list of dependencies.

Julia Version 1.8.5
Commit 17cfb8e65ea (2023-01-08 06:45 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 20 × Intel(R) Core(TM) i9-10900 CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 1 on 20 virtual cores
Environment:
  JULIA_LOAD_PATH = @:@v#.#:@stdlib:/environments/globalenv
  JULIA_DEPOT_PATH = /home/zez25zuh/.julia:
Official https://julialang.org/ release

...
```

**Additional context**
This is neither important nor urgent for me -- it arose originally as a problem in CoCoA.
It could be that no faster way exists.

**Comment by fingolfin** [comment]
This essentially boils down to an invocation of `fmpz_poly_gcd` from FLINT, so I guess any improvements would need to be made there?

Naively, it would of course be no problem to "pull out" all factors `x` from the arguments `f` and `g`, then return `gcd(f/x^a, g/x^b) * x^min(a,b)`. That could even be done as a preprocessing step in Julia. What I don't know, though, is how useful this would in general. I.e. maybe there are cases that get substantially worse by this? Also, I have no idea if perhaps there is a "deeper" issue in the FLINT gcd code that ought to be resolved and which would also make this computation faster?

**Comment by fieker** [comment]
A couple of things - all need or should be done in flint
 - one should remove valuation (divide by x) as this is cheap
 - the modular algorithm computes a heuristic boud for the result - and only tries division once it is past this bound.
   This strategy is bad..

**Comment by JohnAAbbott** [comment]
Yes, removing integer content and power-product content is cheap, so should be done.

OK, we can change the example so that the gcd is `x^2+x-1`.  I imagine that the computation times will remain essentially unaltered.

**Comment by fingolfin** [comment]
I cannot reproduce this (anymore), perhaps some FLINT updated helped? At when I enter the original input (which I slightly tweaked just now to make it work again) I get 
```
julia> @time h = gcd(f,g);
  0.152780 seconds (24.67 k allocations: 58.202 MiB)
```

The by far slowest part is computing `f` and `g`. For the the given input with UPB=100, this takes maybe 30 seconds

**Comment by JohnAAbbott** [comment]
I confirm that for `UPB=100` the final `gcd` computation is now acceptably quick (0.2s on my computer), and certainly far quicker than 100s originally reported.  For comparison, I repeated the computation with initial values for `f` and `g` being `x^2+x-1` (so there is no factor of `x` to divide out): the gcd time was about 45s on my computer -- not sure if this could sensibly be made faster (nor whether it is important to do so).
I'm waiting for the case `UPB=200` to finish generating the input polynomials `f` and `g`; my expectation is that the `gcd` computation will again be quick.  If so, I'll confirm and close this issue.

**Comment by JohnAAbbott** [comment]
The case `UPB=200` took about 5 seconds to compute the gcd (and a lot longer to compute `f` and `g` -- of degrees roughly 24500 and 49000 respectively).  This is fine -->  **Closing**
